Update/34885 category field in product editor

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This updates the product category metabox with the current product editor with a react version of it that replaced the list of categories with a async search dropdown.
The async search dropdown makes use of the Tree Select Control component.

Closes #34885  .

https://user-images.githubusercontent.com/2240960/220586001-39acc0f1-8adf-4170-9050-0447746548db.mp4

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested. Please, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Enable the `async-product-editor-category-field` feature flag using the WooCommerce Beta Tester.
2. Load this branch and create a new store and be sure to create some new Categories.
3. Go to **Products > Add New**
4. The usual category metabox should be present if you have less then a 100 categories.
5. Create either more then a 100 categories or add this filter to your `mu-plugin` -> `add_filter( 'woocommerce_product_category_metabox_search_threshold', function() { return 5; } );`
6. Add a name and scroll down to the Product Categories metabox on the left hand side
7. Try search for some of the categories you previously added and select these.
8. Save the product, the categories should persist.
9. Other items to check:
     - Selecting a popular category should automatically select the category within the popular tab and vice versa
     - Creating a new category should automatically add the new item to the selected list.
     - Creating a new category and selecting a parent should work correctly.

<!-- End testing instructions -->

### Other information:

-   [] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add initial custom meta box for product categories

louwie17 committed:
Make use of TreeSelectControl

louwie17 committed:
Update classnames

louwie17 committed:
Display selected items and sync with most used tab

louwie17 committed:
Always show placeholder and remove checklist container

louwie17 committed:
Reactify category metabox tabs

louwie17 committed:
Add create new category logic

louwie17 committed:
Remove unused markup

louwie17 committed:
Fix saving of empty category list

louwie17 committed:
Add callback when input is cleared as well

louwie17 committed:
Some small cleanup and refactoring.

louwie17 committed:
Add changelog

louwie17 committed:
Fix tree creation and style enqueue

louwie17 committed:
Auto fix lint errors

louwie17 committed:
Fix linting errors

louwie17 committed:
Fix css lint errors

nathanss commented on packages/js/components/src/tree-select-control/options.js:
```diff
@@ -76,6 +78,7 @@ const Options = ( {
 							) }
 							tabIndex="-1"
 							onClick={ ( e ) => {
+								e.preventDefault();
```
Just to confirm: this component is still not used in other places? Why did you had to add this?

louwie17 replied:
I am not sure how this was working in the original Google Listings plugin, but without this the dropdown would auto close and also cause some weird behaviour with losing focus.

nathanss commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+				/>
+			</div>
+			<ul
+				className="categorychecklist form-no-clear tagchecklist"
```
I still get confused about our CSS conventions, but isn't something missing from `categorychecklist` and `tagchecklist`?

louwie17 replied:
Yeah, I wouldn't normally use these, but these were the same classes used in the original metabox. And to keep the same styling for the popular tab and the selected categories (like tags) I kept these classes here.

mattsherman replied:
Maybe a comment in the code would be useful so that future devs understand that this is to use the same styling as the popular tab?

nathanss commented:
I saw these two errors at the console at some point: (trying to figure out how to reproduce, I'll edit if/when I figure it out):

`Warning: validateDOMNesting(...): <ul> cannot appear as a descendant of <p>.`
`Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>.`

nathanss commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-handlers.js:
```diff
@@ -0,0 +1,25 @@
+	if ( container ) {
+		const selectedCategories = Array.from(
+			container.querySelectorAll( ':scope > input[type=hidden]' )
+		).map( ( ele ) => {
```
'ele' is 'him' in Portuguese... That's funny. No need to change it, though 😂

mattsherman replied:
`element` would be more readable. Or even `categoryElement`.

nathanss commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-metabox.tsx:
```diff
@@ -0,0 +1,107 @@
+} > = ( { initialSelected } ) => {
+	const [ selected, setSelected ] = useState( initialSelected );
+	const [ activeTab, setActiveTab ] = useState(
+		initialTab === 'pop' ? 'pop' : 'all'
```
Nitpick: the pattern I usually see is `initialTab === 'pop' ? initialTab : 'all'`

nathanss replied:
Also, are we controlling `product_cat_tab`? It took me some time to realize that `pop` probably stands for `popular` which is the 'Most used' tab. (Is that correct?) I would love if we could be more explicit in this.

louwie17 replied:
I end up replacing its use with a constant, I kept `pop` though to keep inline with the original implementation.

nathanss commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/popular-category-list.tsx:
```diff
@@ -0,0 +1,99 @@
+				<li key={ cat.term_id } className="popular-category">
+					<label
+						className="selectit"
+						htmlFor={
```
Can you assign that string concatenation to a constant since it's used two times? And maybe use the `in-popular-${CATEGORY_TERM_NAME}-${cat.term_id}` syntax?

nathanss commented on plugins/woocommerce/client/legacy/js/admin/wc-enhanced-select.js:
```diff
@@ -305,6 +305,81 @@ jQuery( function( $ ) {
+						}
+					}, getEnhancedSelectFormatString() );
+
+					$( this ).selectWoo( select2_args ).addClass( 'enhanced' );
```
I got totally lost reading this 😳

Can you describe what this does? Maybe increment the comment on line 322 to briefly explain what this is doing.

I know there's similar code on the file, I'm just providing my perspective as someone who is not that experienced in jQuery and its libraries.

louwie17 replied:
yeah this seemingly was left behind from an old implementation and not necessary anymore.

louwie17 replied:
Yeah I removed this, it was left over from a previous implementation.

nathanss reviewed (changes requested):
Overall feedback:

It's working well, I couldn't find any bugs while testing, besides the console error I reported in the previous comment. Congratulations, it must have been pretty hard and complex to do this!

As for the appearance, the sizing looks bigger than the rest of the elements. f.e the 30px height of the input compared to the product tags input, 20px height of the checkbox when comparing to the brand's checkbox. Some screenshots for reference:
<img width="283" alt="image" src="https://user-images.githubusercontent.com/13437655/222441339-cee0182f-fb54-4385-aa62-0e219c3bd9f8.png">
<img width="299" alt="image" src="https://user-images.githubusercontent.com/13437655/222441388-3ba0138e-34a2-4546-9650-9954ca43ec78.png">

louwie17 committed:
Add 100 limit, and address some PR feedback

louwie17 committed:
Fix some styling and warnings

louwie17 committed:
Remove unused code

louwie17 committed:
Address PR feedback

louwie17 committed:
Fix lint error

louwie17 commented:
@nathanss this should be ready for a re-review, I addressed all your changes, thanks for the initial review and sorry for the delay in updating this :)

louwie17 committed:
Fix lint errors

mattsherman commented on plugins/woocommerce/changelog/update-34885_category_field_in_product_editor:
```diff
@@ -0,0 +1,4 @@
+Significance: major
+Type: update
+
+Update Category product metabox with a async dropdown search control rendered with React.
```
`a async` => `an async`

mattsherman commented on plugins/woocommerce/includes/admin/class-wc-admin-assets.php:
```diff
@@ -194,9 +195,26 @@ public function admin_scripts() {
+					WCAdminAssets::get_url( "product-category-metabox/style{$rtl}", 'css' ),
+					array( 'wp-components' ),
+					WCAdminAssets::get_file_version( 'css' )
+				);
```
Why not put this in `admin_styles()` above?

And, why is the RTL handling done differently than in `admin_styles()`?

mattsherman commented on plugins/woocommerce/includes/admin/class-wc-admin-assets.php:
```diff
@@ -194,9 +195,26 @@ public function admin_scripts() {
+						'search_categories_nonce'     => wp_create_nonce( 'search-categories' ),
+						'search_taxonomy_terms_nonce' => wp_create_nonce( 'search-taxonomy-terms' ),
+					)
+				);
```
This feels a little odd, since it doesn't seem to have anything to do with localization. Why is this needed here?

mattsherman commented on plugins/woocommerce/includes/class-wc-ajax.php:
```diff
@@ -1764,12 +1765,14 @@ public static function json_search_categories() {
 		}
 
+		$hide_empty       = isset( $_GET['hide_empty'] ) ? wc_clean( wp_unslash( $_GET['hide_empty'] ) ) : true;
+		$hide_empty       = 'false' === $hide_empty ? false : true;
```
I think you can simplify this to...

```php
$hide_empty = isset( $_GET['hide_empty'] ) ? wp_validate_boolean( $_GET['hide_empty'] ) : true;
```

But, the negative flag still bugs me... did you consider going with a flag `show_empty` and defaulting that to `false`? That seems to be a bit easier to comprehend to me.

mattsherman commented on plugins/woocommerce/includes/class-wc-ajax.php:
```diff
@@ -1798,6 +1803,76 @@ public static function json_search_categories() {
+		}
+
+		$search_text = isset( $_GET['term'] ) ? wc_clean( wp_unslash( $_GET['term'] ) ) : '';
+		$number      = isset( $_GET['number'] ) ? absint( wp_unslash( $_GET['number'] ) ) : 50;
```
Why would the `number` query param ever have slashes in it? Calling `wp_unslash()` feels unnecessary.

mattsherman commented on plugins/woocommerce/includes/class-wc-ajax.php:
```diff
@@ -1798,6 +1803,76 @@ public static function json_search_categories() {
+		if ( $terms ) {
+			foreach ( $terms as $term ) {
+				$terms_map[ $term->term_id ] = $term;
+				$term->formatted_name        = '';
```
Why is `$term->formatted_name` set to `''` and then never set to anything else? This feels like a bug. If it is not, it might deserve a comment, as it definitely invites questions!

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+		};
+		categoryLibrary[ child.term_id ] = child;
+		newTree.push( newItem );
+		if ( child.children && child.children.length > 0 ) {
```
You could do `child.children?.length > 0`.

louwie17 replied:
Thanks, done :)

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+
+	useEffect( () => {
+		apiFetch< CategoryTreeItem[] >( {
+			url: addQueryArgs( getSetting( 'adminUrl' ) + 'admin-ajax.php', {
```
`addQueryArgs( new URL( 'admin-ajax.php', getSettings( 'adminUrl' ) ).toString(), {} )` would be a bit more robust, in case `adminUrl` didn't have a slash at the end.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+				<TreeSelectControl
+					alwaysShowPlaceholder={ true }
+					options={ treeItems }
+					value={ selected.map( ( cat ) => cat.term_id.toString() ) }
```
`category` would be more descriptive than `cat`. And, better for searching.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+					value={ selected.map( ( cat ) => cat.term_id.toString() ) }
+					onChange={ ( sel: number[] ) => {
+						onChange( sel.map( ( id ) => categoryLibrary[ id ] ) );
+					} }
```
`sel` isn't very descriptive. Maybe something like `selectedCategoryIds` would be more readable?

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+}
+
+export const AllCategoryList: React.FC< {
+	selected: CategoryTerm[];
```
Maybe `selectedCategoryTerms`?

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+				className="categorychecklist form-no-clear tagchecklist"
+				id={ CATEGORY_TERM_NAME + 'checklist' }
+			>
+				{ selected.map( ( cat ) => (
```
Similar comment about `cat` as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/all-category-list.tsx:
```diff
@@ -0,0 +1,127 @@
+							className="ntdelbutton"
+							onClick={ () => {
+								const newSelectedItems = selected.filter(
+									( sel ) => sel.term_id !== cat.term_id
```
Similar comment about `sel` as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-add-new.tsx:
```diff
@@ -0,0 +1,160 @@
+};
+
+export const CategoryAddNew: React.FC< {
+	selected: CategoryTerm[];
```
Similar comment about `selected` as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-add-new.tsx:
```diff
@@ -0,0 +1,160 @@
+
+		const data = {
+			name: newCategoryName,
+			parent: categoryParent ? categoryParent.term_id : -1,
```
Consider `categoryParent?.term_id ?? -1`

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-add-new.tsx:
```diff
@@ -0,0 +1,160 @@
+			setFetchedItems( [] );
+			return apiFetch< CategoryTerm[] >( {
+				url: addQueryArgs(
+					getSetting( 'adminUrl' ) + 'admin-ajax.php',
```
Same comment as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-add-new.tsx:
```diff
@@ -0,0 +1,160 @@
+					}
+				),
+				method: 'GET',
+			} ).then( ( res ) => {
```
Consider `response` instead of `res`

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-add-new.tsx:
```diff
@@ -0,0 +1,160 @@
+				className="taxonomy-add-new"
+				onClick={ () => setShowAddNew( ! showAddNew ) }
+			>
+				{ __( '+ Add new category', 'woocommerce' ) }
```
The plus at the beginning of this string doesn't seem ideal for screen readers. Maybe we should have an `aria-label` attribute on the `<a>` set to `Add new category`?

louwie17 replied:
Good idea

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/category-metabox.tsx:
```diff
@@ -0,0 +1,142 @@
+	initialTab = window.getUserSetting( CATEGORY_TERM_NAME + '_tab' ) || '';
+}
+
+const CATEGORY_POPULAR_TAB_ID = 'pop';
+const CATEGORY_ALL_TAB_ID = 'all';
```
More of a question vs a request for a change... do we have any specific guideline that says whether to do ALL-CAPS for constants?

louwie17 replied:
I don't think we have a specific guideline, or at-least that I know off.
I guess I have always been used to capping all my constants.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/popular-category-list.tsx:
```diff
@@ -0,0 +1,91 @@
+
+	useEffect( () => {
+		apiFetch< CategoryTerm[] >( {
+			url: addQueryArgs( getSetting( 'adminUrl' ) + 'admin-ajax.php', {
```
Same comment as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-category-metabox/style.scss:
```diff
@@ -0,0 +1,37 @@
+.product-add-category {
+	&__tree-control {
+		margin-top: 8px;
```
Do we have access to any SCSS variables so that we don't have to hardcode values here?

louwie17 replied:
Yup added where possible.

mattsherman commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-categories.php:
```diff
@@ -0,0 +1,73 @@
+		$tax_name            = $parsed_args['taxonomy'];
+		$selected_categories = wp_get_object_terms( $post->ID, 'product_cat' );
+		?>
+		<div id="taxonomy-<?php echo esc_attr( $tax_name ); ?>-metabox"></div>
+		<?php foreach ( (array) $selected_categories as $term ) { ?>
+			<input
+				type="hidden"
+				value="<?php echo esc_attr( $term->term_id ); ?>"
+				name="tax_input[<?php echo esc_attr( $tax_name ); ?>][]"
+				data-name="<?php echo esc_attr( $term->name ); ?>"
```
Could pull `esc_attr( $tax_name )` out into a variable outside of the `foreach` loop.

louwie17 replied:
Yeah I believe I tried that prior, but then I get a `EscapeOutput.OutputNotEscaped` lint error even though the variable is clearly escaped.

mattsherman reviewed (changes requested):
Overall, works pretty well, though I didn't test extensively, since I spent a lot of time reviewing the code (this PR would have been nice to split up... perhaps into one that just reviewed the React component and another that focused on the integration).

@louwie17 This is a pretty impressive job... it's cool to see the existing screen being evolved with a React component. 🥳 

Left a bunch of comments with suggestions for readability improvements.... most of them are just suggestions.

Functionality-wise, one change that jumped out to me was that you can't select a child without selecting the parent in the new tree. You can remove the parent from the list below the "Add category" field though, so maybe that is okay?

Echoing what @nathanss already mentioned, the font size in the new component looks too big compared with the rest of the UI it is in...

<img width="303" alt="Screenshot 2023-03-27 at 20 22 05" src="https://user-images.githubusercontent.com/2098816/228096508-b7875b35-1710-4a58-a9de-55b110e79104.png">

Changing the font size to `12px` results in a things looking better, though we'd also want to adjust the checkbox sizing since it still looks too big...

<img width="303" alt="Screenshot 2023-03-27 at 20 24 40" src="https://user-images.githubusercontent.com/2098816/228096642-a7f7d6a6-e3d7-4f35-9bb5-396693b8cc8a.png">

The case where a user enters in text into the "Add category" field that doesn't match anything feels off... I'd probably expect to be able to add a new category when I did that, instead of the field just sitting there with no indication that anything is wrong or what to do next.

louwie17 committed:
Address PR feedback

louwie17 committed:
Fix lint error

louwie17 committed:
Minor fixes and add tracking

louwie17 committed:
Add debounce

louwie17 committed:
Fix lint error

louwie17 commented:
@mattsherman thanks for the detailed review, I believe I have addressed all your feedback and fixed the font sizing issues.

This PR should be ready for a re-review.

mattsherman commented:
> @mattsherman thanks for the detailed review, I believe I have addressed all your feedback and fixed the font sizing issues.
> 
> This PR should be ready for a re-review.

There are a number of unresolved comments from the last time I reviewed this PR. I have went through and marked the ones that were actually resolved as resolved. It would be great to resolve those remaining unresolved comments, ideally with a short comment, so that I know you looked at the comment and considered it. Many are suggestions, so it's okay if you resolve them by saying "thanks but no thanks" 😄. I just don't want them to go unnoticed!

More importantly though, there are a number of bugs I found in my latest round of testing...

- Add category field
  -  [ ] If the dropdown is dismissed via ESC key, it doesn't come back unless you un-focus and then re-focus the field
  -  [ ] When I start typing "ho", "Clothing", the parent for "Hoodies" shows in the dropdown list. But, "Hoodies" itself doesn't show until I type "hoo"
- Parent category search
  - [ ] When typing something in, immediately erases it
  - [ ] Once a parent is selected from dropdown list, there is no way to remove the parent category
  - [ ] No way to select text in the parent category field

Here is a video that demos the bugs... complete with a non-polished voice-over!

https://user-images.githubusercontent.com/2098816/231619382-16e6a802-40bc-4669-be32-8c36a216b3ef.mov

mattsherman reviewed (changes requested):
Found bugs that need to be addressed (see other comment).

louwie17 committed:
Allow custom min filter amount and fix menu not showing after escaping input

louwie17 committed:
Allow single item to be cleared out of select control

louwie17 committed:
Fix bug where typed values did not show up

louwie17 committed:
Fix some styling issues

louwie17 committed:
Allow parents to be individually selected

louwie17 committed:
Address PR feedback and add error message

louwie17 committed:
Add changelogs

louwie17 commented:
Thanks for taking another detailed look @mattsherman 💯 

> There are a number of unresolved comments from the last time I reviewed this PR. I have went through and marked the ones that were actually resolved as resolved. It would be great to resolve those remaining unresolved comments, ideally with a short comment, so that I know you looked at the comment and considered it. Many are suggestions, so it's okay if you resolve them by saying "thanks but no thanks" 😄. I just don't want them to go unnoticed!

Sorry, and thanks for calling this out. I felled in a bit of a hurry (for some reason, I think before my AFK at that time), so end up only replying to a few. I believe I have addressed all your items now, and left a couple comments or responded with an Emoji :)

> * Add category field
>   
>   * [ ]  If the dropdown is dismissed via ESC key, it doesn't come back unless you un-focus and then re-focus the field

This was a bug with the Tree select control, I have fixed it here: https://github.com/woocommerce/woocommerce/pull/36869/commits/6729f4d2598ecdab9a380735a2791e19a13b427a ( I fixed the item below in this commit as well)

>   * [ ]  When I start typing "ho", "Clothing", the parent for "Hoodies" shows in the dropdown list. But, "Hoodies" itself doesn't show until I type "hoo"

Yeah they had a minimum filter limit set, which was set to 3, I added a prop to manually set it and changed it to `2` (1 seemed a bit unnecessary).

> * Parent category search
>   
>   * [ ]  When typing something in, immediately erases it

This was an annoying one to figure out, but this was because of a `useEffect` logic that had the `getItemLabel` as a dependency, moving this function outside of the component fixed this: https://github.com/woocommerce/woocommerce/pull/36869/commits/b0d4568a8d973edd60cd79448c8e8207b3e0275f

>   * [ ]  Once a parent is selected from dropdown list, there is no way to remove the parent category

I added support for this in the tree select control by exposing a `individuallySelectParent` prop: https://github.com/woocommerce/woocommerce/pull/36869/commits/d20757384ed114b98943849ab0cd6da75aa04496

>   * [ ]  No way to select text in the parent category field

This should be fixed now, this was a bug in the select control where `preventDefault` was being called even when the `input` field was being clicked, fixed here: https://github.com/woocommerce/woocommerce/pull/36869/commits/4251ebe141a45cd8c7726c043bd703087f2bc2d7

@mattsherman this should be ready for a re-view assuming all the actions will pass at the time of writing this :)

mattsherman commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-categories.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Product Images
+ *
+ * Display the product images meta box.
```
Copy pasta :-) Didn't notice this before.

louwie17 replied:
Nice catch

mattsherman commented on plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-categories.php:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * WC_Meta_Box_Product_Images Class.
+ */
```
More copy pasta.

mattsherman commented on plugins/woocommerce/includes/class-wc-post-types.php:
```diff
@@ -123,6 +123,7 @@ public static function register_taxonomies() {
 						'with_front'   => false,
 						'hierarchical' => true,
 					),
+					'meta_box_cb'           => 'WC_Meta_Box_Product_Categories::output',
```
@louwie17 I'm seeing really odd behavior...

With this set as such, saving does not work properly, with either the new typeahead field, or the existing field. It seems to be using the category `name` as the `term_id` when saving?!

If I change this to...

```
'meta_box_cb' => 'post_categories_meta_box'
```

Then things work properly (it always uses the old field, of course).

I tried debugging this a bit, but I can't figure out what is happening!

mattsherman replied:
`WC_Meta_Box_Product_Categories::output` is being called, and the correct params seem to be passed. And, the logic correctly called `post_categories_meta_box` in there. But, it acts differently.

mattsherman replied:
@louwie17 This bug was really haunting me... I couldn't get it out of my brain and needed to understand what was going on... I dug more into the underlying WP code and determined the issue...

We need to set 

```
'meta_box_sanitize_cb' => 'taxonomy_meta_box_sanitize_cb_checkboxes',
```

Because, without doing this, `meta_box_sanitize_cb` defaults to "an appropriate one based on the value of `$meta_box_cb`" ([doc](https://developer.wordpress.org/reference/functions/register_taxonomy/)), which ends up being `taxonomy_meta_box_sanitize_cb_input` for any `$meta_box_cb` that isn't `post_categories_meta_box` ([src](https://github.com/WordPress/wordpress-develop/blob/e0005a5788630717740ccb7150ae39d0feeae789/src/wp-includes/class-wp-taxonomy.php#L452)).

I tested this change with both the existing category field and the new typeahead field and things appear to function properly now!

Please make this change and test it youself to verify that the save works as intended, and then let me know and I'll give this a re-re-re-review. 😄

mattsherman commented:
@louwie17 In addition to the **blocking bug of the save not working properly** (see comment previously), there are some more minor UX details I've noticed...

- [ ] I'm not sure what order the categories are being shown in
- [ ] If I just click on the field and scroll through the dropdown, it is not clear to me that all categories are not being shown
- [ ] Using keyboard to select child item after doing a search results in confusing display where the parent is collapsed -- it doesn't look like the child was selected
- [ ] I don't know if it is "right" or not, but I really want to press RETURN to select items in dropdown
- [ ] I want to see child items when I search for the parent... maybe I don't remember all the child choices, but I do remember the parent; or, maybe I don't realize there are child choices

https://user-images.githubusercontent.com/2098816/231879511-8d075655-48fa-4111-8121-0357512126b7.mov

These can probably wait until a future release/PR, but they would be great to get in there at some point. /cc @jarekmorawski @octaedro

louwie17 committed:
Fix saving issue

louwie17 committed:
Add client side sorting and stop clearing field upon selection

louwie17 committed:
Update changelog

louwie17 commented:
> @louwie17 In addition to the **blocking bug of the save not working properly** (see comment previously), there are some more minor UX details I've noticed...

Thanks for looking into the fix for this, that made my job a lot easier :)

> * [x]  I'm not sure what order the categories are being shown in

I defaulted to `name` sorted on the client side.

> * [ ]  If I just click on the field and scroll through the dropdown, it is not clear to me that all categories are not being shown

I skipped this one, as it would require some design input and I am not entirely sure what the best approach/best practice here is. Could be a follow up though.

> * [x]  Using keyboard to select child item after doing a search results in confusing display where the parent is collapsed -- it doesn't look like the child was selected

Yeah it clears the input field upon selection, I added a `prop` to disable this, so things should feel a bit more normal now.

> * [ ]  I don't know if it is "right" or not, but I really want to press RETURN to select items in dropdown
> * [ ]  I want to see child items when I search for the parent... maybe I don't remember all the child choices, but I do remember the parent; or, maybe I don't realize there are child choices

I am not sure about these two either, the parent auto expanded is definitely a preference thing, I believe we have the same logic in the new UI with the category field (on not auto expanding unless a child is matched).
We may also want to switch the component to the tree select control that Nathan worked on, which may use RETURN in this way, I am not a 100% sure.

jarekmorawski commented:
I agree that this needs a bit more polish before the launch. 🚀 In addition to fixing the UX issues Matt listed, I think we can improve the UI to make it clearer and more scalable. 

> I'm not sure what order the categories are being shown in

We should order the categories alphabetically, both parents and their children.

>  If I just click on the field and scroll through the dropdown, it is not clear to me that all categories are not being shown

Do we know why? Is a technical constraint preventing us from loading all parent categories and their children? Do you know if we need to add a button to load the remaining items? The way I see it, we could load the first 20-30 items (both after the user clicks the field and while searching) with all parents collapsed. We would load the rest async as the user scrolls down the list and only load the children (all, if possible) when the user clicks the expand arrow.

If needed, we could show a loading spinner in place of the chevron while the items are loaded.

>  I don't know if it is "right" or not, but I really want to press RETURN to select items in dropdown

Good idea. Both `Space` and `Return/Enter` are acceptable regarding accessibility. However, for parent items, `Return` should open/close them, not select. 

>  I want to see child items when I search for the parent... maybe I don't remember all the child choices, but I do remember the parent; or, maybe I don't realize there are child choices

That's reasonable, especially when you may want to select just a few child items but not the parent. It could be collapsed by default. We'd only auto-expand when there's a match for a child item (so that a parent always accompanies it).

In addition to the above:

- [ ] Child items should have an indent the size of the parent's box.
- [ ] Users should be able to select a child or a parent separately.
- [ ] Whenever a child is selected, the parent's box should gain the mixed state (empty box with a horizontal line). This should apply to search results and the list in a resting state.
- [ ] We don't need the horizontal line separating the parent from the children.
- [ ] When the search field is focused, it should always be interactive, allowing users to enter or update the query even if their cursor is elsewhere.
- [ ] It's okay to decrease spacings a bit to save some space. Can we reduce the paddings on the sides and the distance between the items? Using the smaller version of the chevron icon would also help.
- [ ] The search field's placeholder text should be `Type to search` to avoid confusion with the `Add new category` button below.
- [ ] We can rename the tab from `All items` to `All categories`.

louwie17 commented:
> Do we know why? Is a technical constraint preventing us from loading all parent categories and their children? Do you know if we need to add a button to load the remaining items? The way I see it, we could load the first 20-30 items (both after the user clicks the field and while searching) with all parents collapsed. We would load the rest async as the user scrolls down the list and only load the children (all, if possible) when the user clicks the expand arrow.

I currently have it limited to displaying 50 items. Auto loading the remaining items as a user scrolls is kinda tricky with our current data structure, as we are displaying it in a tree like structure, but the data is saved in a single list in the database. This makes the pagination around this rather difficult.

Maybe we could just add an item that mentions if you want to few more items to start searching to narrow it down? (or something to that affect). As we are already querying the database when a user types. I do think we want to encourage a user typing to search for a category versus scrolling down the list ( which the old UI was already doing ).
I can also up the limit a bit, to a 100, or 200.

louwie17 committed:
Create feature flag for async product categories dropdown

louwie17 commented:
@mattsherman I wrapped the logic in a `async-product-editor-category-field` feature flag, so it will only modify the category taxonomy config type if the feature is enabled.
This will allow us to merge this first part that has been tested, and to work on the follow ups.

louwie17 committed:
Fix lint errors

louwie17 committed:
Fix linting

mattsherman reviewed (approved):
@louwie17 Thanks for putting this behind a feature flag.

I verified that with the feature flag disabled, the old category selector is used. With the feature flag enabled, the new category selector is used if the number of categories exceeds the threshold.
